processing/src/main/java/io/druid/segment/DimensionIndexer.java

```diff
@@ -0,0 +1,323 @@
+ *
+ * 1.) Ordering based on encoded value's order of ingestion
+ * 2.) Ordering based on converted actual value
+ *
+ * Suppose we have a new String dimension DimA, which sees the values "Hello", "World", and "Apple", in that order.
+ * This would correspond to dictionary encodings of "Hello"=0, "World"=1, and "Apple"=2, by the order
+ * in which these values were first seen during ingestion.
+ *
+ * However, some use cases require the encodings to be sorted by their associated actual values.
+ * In this example, that ordering would be "Apple"=0, "Hello"=1, "World"=2.
+ *
+ * The first ordering will be referred to as "Unsorted" in the documentation for this interface, and
+ * the second ordering will be referred to as "Sorted".
+ *
+ * @param <EncodedType> class of the encoded values
+ * @param <ActualType> class of the actual values
+ *
+ */
+public interface DimensionIndexer<EncodedType extends Comparable<EncodedType>, ActualType extends Comparable<ActualType>>
+{
+  /**
+   * Given a single row value, update any internal data structures with the ingested values and
+   * return the row values as an array to be used as a TimeAndDims key.
+   *
+   * For example, the dictionary-encoded String-type column will return an int[] containing a dictionary ID.
+   *
+   * The value within the returned array should be encoded if applicable, i.e. as instances of EncodedType.
+   *
+   * NOTE: This function can change the internal state of the DimensionIndexer.
+   *
+   * @param dimValues Single row val to process
+   *
+   * @return An array containing an encoded representation of the input row value.
+   */
+  public Object processSingleRowValToIndexKey(Object dimValues);
+
+
+  /**
+   * Given a list of row values (for multivalued dimensions), update any internal data structures with the ingested
+   * values and return the row values as an array to be used as a TimeAndDims key.
+   *
+   * For example, the dictionary-encoded String-type column will return an int[] containing dictionary IDs.
+   *
+   * The values within the returned array should be encoded if applicable, i.e. as instances of EncodedType.
+   *
+   * NOTE: This function can change the internal state of the DimensionIndexer.
+   *
+   * @param dimValuesList List of row values to process
+   * @return An array containing encoded representations of the input row values
+   */
+  public Object processRowValsListToIndexKey(Object dimValuesList);
+
+
+  /**
+   * Given an encoded value that was ordered by time of ingestion, return the equivalent
+   * encoded value ordered by associated actual value.
+   *
+   * Using the example in the class description:
+   *   getSortedEncodedValueFromUnsorted(0) would return 2
+   *
+   * @param unsortedIntermediateValue value to convert
+   * @return converted value
+   */
+  public EncodedType getSortedEncodedValueFromUnsorted(EncodedType unsortedIntermediateValue);
```

Comment:
mm, nevermind, I guess it does work since `sortedLookup` is invalid. But, the sorted ids could change. So _that_ should be part of the contract (i.e. if you do getSorted -> processSingleRow -> getSorted again, then you might not get the same id back)